### ISSUE TITLE
Spotty Checker

### ISSUE THREAD

**Issue by jjones6443** [validator/validator#724]
About every other time I check the SAME document it will say it is wrong every other time it says it is correct. Please fix

**Comment by sideshowbarker** [comment]
Please provide the URL for the document or else a copy of the document that I can test with

**Comment by sideshowbarker** [comment]
It’d also help if you indicate exactly what errors the checker is reporting

**Comment by sideshowbarker** [comment]
Closing due to lack of response

**Comment by DamonHD** [comment]
I'm having quite a lot of trouble with vnu.jar at the moment.

I had been using it to batch check validity of a whole set of html files in one go, see the script at:

    http://www.earth.org.uk/script/vnu

This was starting to report spurious errors allegedly in a particular file X (often the first one) as part of a batch of ~200 files, but X is in fact entirely fine, and passes without complaint when checked on its own.

No other indication of distress such as OOE is given.

I've currently switched to checking the files individually, but that costs ~30s of startup time which is pretty horrible per published HTML file rebuild!

It's the sort of behaviour I'd expect from a C/C++ program with pointer/heap troubles, not Java!

Rgds

Damon

**Comment by sideshowbarker** [comment]
Based just on the information provided so far in https://github.com/validator/validator/issues/724#issuecomment-450673020, I’m not able to reproduce the problem described there (*“spurious errors allegedly in a particular file X (often the first one) as part of a batch of ~200 files”*).

To help isolate the cause, it would help if you provide some additional information:

* What exactly are the spurious errors that get reported? Is it the same error every time, at the same line number?
* If you try giving vnu.jar a batch of fewer files, can you still reproduce the problem? What happens if, say, you give it 50 files instead? Or 10 files?

One thing that’s not clear to me is exactly how you’re feeding files to your script. As far as I can tell, the script seems to expect a list of files. So if you’re checking a batch of 200 files, then the arguments you’re giving the script are that list of all 200 files together?

If that’s what you’re doing, why? Are the files spread across different directories or something? Because if, say, you had all the files within some particular directory tree, then you could just give vnu.jar the path to the top of that directory tree, and it will on its own recurse through the tree and check all the files it finds (by default it’ll check any file regardless of extension, but if you want it to only check HTML files, you can use the `--skip-non-html` option to make it do that).

Anyway, I don’t recommend feeding a single list of 200 file/pathnames at once to vnu.jar — because when you’re giving an argument list that long, I’m not confident you won’t end up running into problems actually caused by the shell truncating something, or else even some internal java length limit or something. 

> I had been using it to batch check validity of a whole set of html files in one go, see the script at:
> 
> ```
> http://www.earth.org.uk/script/vnu
> ```

One thing I notice in that script is that it’s setting `-Xmx256m`.  So another question:

* Can you reproduce the problem when you omit that Java option?
* Can you reproduce the problem when you set higher values for that option? (e.g., 512m or 1024m)

Running vnu.jar under a limit of just 256MB isn’t recommended. I’d expect that would cause an OOME at some point — but it if doesn’t then I dunno, maybe it’s causing some other kind of limit to get hit that results in some other kind of subtle internal error that’s not getting reported.

**Comment by sideshowbarker** [comment]
>     1. The error reported is usually about unexpected characters before the DOCTYPE, but there aren't any.

OK, yeah, so that’s specifically a parser error — which happens early (first) in the checking process, before any of the actual validation — and it would seem to suggest some spurious bytes from somewhere are somehow ending up in the stream that the HTML parser sees.

To troubleshoot this, it’d help a lot if you could you please try:

1. Call vnu.jar with either `--format json` or `--format xml`
2. In the JSON or XML error output for the case where you get an “unexpected characters before the DOCTYPE” error, check the value of the `extract` field (JSON) or element (XML). That should show the exact characters the parser is seeing before the doctype.

It’s possible that the bytes that are ending up before the doctype might be non-printing ones that aren’t going to show up in the console, so you may want to redirect the output (from stderr) to a file. And then I could look at it with a hex editor or whatever (or you could yourself too) to see what the actual bytes are — which may or may not give a clue as to where they might be coming from.

**Comment by DamonHD** [comment]
I'll try to test this this this evening.

Do you mean test against .json or .xml files while using the appropriate flag?  Do they have to be large, or just lots of them, or with long filenames in aggregate?

Rgds

Damon

**Comment by sideshowbarker** [comment]
> Do you mean test against .json or .xml files while using the appropriate flag?

No, I mean just against the exact same set of HTML files you’ve already been testing with (the 200 or however many HTML files you mentioned in earlier comments) — but with the error output formatted as JSON or XML, instead of in the default (GNU) error-output format.

(To be clear, the `--format` option controls the format of the error output; so it’s unrelated/orthogonal to the format of the input files you’re checking — which would generally be HTML, SVG, CSS. In hindsight, I realize that giving that option that name `--format` may not have been the best choice as far as making it clear what its purpose is…) 

> Do they have to be large, or just lots of them, or with long filenames in aggregate?

I can’t say that I have any ideas to offer about what particular qualities the files might need to have in order to trigger the problem you’ve reported — the only thing I can suggest is that it should be some facsimile of whatever set of files you were testing against in the past when you ran into the problem.

**Comment by DamonHD** [comment]
Just had a quick play while stuck on something else...

Reducing to max of 32 files and 8192 bytes of command line arg per execution using xargs (and running on macOS) resulted in a failure:

    "file:/Users/dhd/SVN-controlled/www.hd.org/Damon/Env/script/vnu":1.1-16.38: error: Non-space characters found without seeing a doctype first. Expected “<!DOCTYPE html>”.
    "file:/Users/dhd/SVN-controlled/www.hd.org/Damon/Env/script/vnu":1.1-16.38: error: Element “head” is missing a required instance of child element “title”.
    Failed HTML5 validation for off-grid-stats-historical-200909.html...

And, as before. off-grid-stats-historical-200909.html validated on its own is fine.

I will try to to the format stuff when I officially have some time later!  B^>

Rgds

Damon

**Comment by DamonHD** [comment]
Interestingly I'm currently having a hard time making vnu.jar fail if I use the --format option with xml or json!

**Comment by DamonHD** [comment]
So, caught a failure with --format json.

Very odd.   And it's all my fault!  In certain cases Vnu will be fed the script/vnu script as one of its arguments.  I can explain why if you wish.   TL;DR: I am an idiot!

But possibly the error reporting even under gnu format could be improved to indicate correctly that the failure is in a non-html file!

Anyhow, thanks for the steer.  I think that I can fix this now!

Rgds

Damon
